implement the rest of preaggregated metrics using datastax

shintasmith opened the pull request:
This implements the rest of the pre-aggregated metrics IO using datastax driver. Changes are:
* created CounterIO, GaugeIO, SetIO, and TimerIO for Datastax (prefixed with 'D')
* implemented DPreaggregatedMetricsRW
* implemented APreaggregatedMetricsRW for Astyanax driver, it is just wrappers to call the appropriate methods in AstyanaxReader and AstyanaxWriter
* used in LocatorIO when writing metrics
* merged in tests in service.PreaggregatedMetricsIntegrationTest to io.PreaggregatedMetricsRWIntegrationTest

ChandraAddala commented on blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/io/PreaggregatedMetricsRWIntegrationTest.java:
```diff
@@ -0,0 +1,543 @@
+     *
+     * @return
+     */
+    protected Object getGranularitiesToTest() {
```
Why not use `Granularity.granularities()` method?

shintasmith replied:
This method is being called by the junitparams, when you annotate a test method like this:
```
@Parameters(method = "getGranularitiesToTest")
```
The junitparams will then iterate through the array and pass in the granularity as a parameter to the test method. I think it's good to control the input within the test method.

ChandraAddala replied:
Tomorrow if we add a granularity, someone has to remember to add it here also. Can you call `Granularity.granularities()` inside the method? If its not possible thats okay.

shintasmith replied:
Ok. Good point.

ChandraAddala commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DAbstractPreaggregatedIO.java:
```diff
@@ -0,0 +1,333 @@
+ * This is an abstract class that collects the common behavior of
+ * reading/writing preaggregated metrics objects.
+ */
+public abstract class DAbstractPreaggregatedIO {
```
What does "D" signify?

ChandraAddala replied:
Nevermind. Got it.

shintasmith replied:
:-) Sorry, the name was getting too long.

ChandraAddala replied:
`DatastaxAbstractPreaggregatedIO` is not too long. You know what is too long? This guy `SimpleBeanFactoryAwareAspectInstanceFactory`

shintasmith replied:
😱 😂. How about:
DatastaxPreaggregatedMetricsReadWrite ?

ChandraAddala commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DAbstractPreaggregatedIO.java:
```diff
@@ -0,0 +1,333 @@
+    protected PreparedStatement insertToMetricsPreaggr240MStatement;
+    protected PreparedStatement insertToMetricsPreaggr1440MStatement;
+
+    private PreparedStatement selectFromMetricsPreaggrFullForRangeStatement;
```
can we make these final?

shintasmith replied:
They all get initialized in the constructor. How can I make it final above?

ChandraAddala replied:
Something like this.
```
public class Person {
    private final String name;

    public Person(String name) {
        this.name = name;
    }
}
```

shintasmith replied:
Gotcha. For some reasons I thought I would get compile error 😊. I'll make em final.

ChandraAddala commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/service/RollupRunnable.java:
```diff
@@ -103,6 +104,19 @@ public void run() {
+            //              | SimpleNumber (if gran == full) | metrics_full
+            // BF_HISTOGRAM | HistogramRollup                | metrics_full (if gran == full)
+            //              |                                | metrics_preaggr_{gran} otherwise
+
```
:thumbsup:

usnavi commented on blueflood-core/src/integration-test/java/com/rackspacecloud/blueflood/io/PreaggregatedMetricsRWIntegrationTest.java:
```diff
@@ -0,0 +1,543 @@
+            List<SingleRollupWriteContext> writeContexts = toWriteContext(expectedLocatorMetricMap, granularity);
+            datastaxMetricsRW.insertRollups(writeContexts);
+
+            // read with datastax
```
What do you mean by "read with datastax"?

shintasmith replied:
sorry, copy-pasta error, I'll fix it

usnavi commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DAbstractPreaggregatedIO.java:
```diff
@@ -0,0 +1,333 @@
+                List<ResultSetFuture> futures = entry.getValue();
+
+                Table<Locator, Long, T> result = toLocatorTimestampRollup(futures, locator, CassandraModel.getGranularity(columnFamily));
+                locatorTimestampRollup.putAll(result);
```
what might throw and exception in this block?

shintasmith replied:
the ```toLocatorTimestampRollup()``` is where the future.getInterruptibly() get called. Do you think I should move the try-catch inside the toLocatorTimestampRollup(), which currently doesn't do any Instrumentation calls? Maybe I could just declare the exceptions as stated in the Datastax docs?

usnavi replied:
I thought toLocatorTimestampRollup() was catching the exception, which i think is right as it won't stop the loop from processing.

https://github.com/rackerlabs/blueflood/pull/678/files/457f1ef9f225335824b2c9c36be97c25bef0c23c#diff-87684b60fe53b87d4109d19ed05e4dfbR314

shintasmith replied:
You are right. I will remove this try-catch.

ChandraAddala commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/astyanax/APreaggregatedMetricsRW.java:
```diff
@@ -0,0 +1,133 @@
+    public MetricData getDatapointsForRange(final Locator locator,
+                                            Range range,
+                                            Granularity gran) throws IOException {
+        return AstyanaxReader.getInstance().getDatapointsForRange(locator, range, gran);
```
This doesnt seem to throw `IOException`. When will we get IOException?

shintasmith replied:
getDatapointsForRange() method is defined in the MetricsRW interface. Both Astyanax and Datastax implements this. The Datastax will throw IOException. The Astayanx may not. Should I catch a generic RuntimeException here and rethrow as IOException ?

ChandraAddala replied:
No. its fine.

usnavi commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DPreaggregatedMetricsRW.java:
```diff
@@ -0,0 +1,349 @@
+                MetricData metricData = new MetricData(points, getUnitString(locator), outputType);
+                locatorMetricDataMap.put(locator, metricData);
+
+            } catch (Exception ex) {
```
what throws an exception here?

shintasmith replied:
The toLocatorTimestampRollup() does future.getInterruptibly()

shintasmith replied:
I was incorrect, as you point out. The toLocatorTimestampRollup() has its own try-catch. But I will change this ```catch(Exception ex)``` to ```catch(CacheException ex)``` as the ```getDataType()``` throws it.

usnavi commented:
I assume this will be wired into blueflood in a later PR?

shintasmith commented:
Yes, that is the task CMD-1032

ChandraAddala commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/datastax/DCounterIO.java:
```diff
@@ -0,0 +1,60 @@
+        if ( ! (rollup instanceof BluefloodCounterRollup) ) {
+            throw new IllegalArgumentException("toByteBuffer(): expecting BluefloodCounterRollup class but got " + rollup.getClass().getSimpleName());
+        }
+        BluefloodCounterRollup counterRollup = (BluefloodCounterRollup) rollup;
```
`instanceof` check and the casting happens for all classes implementing `DAbstractPreaggregatedIO` Any idea how we can avoid them?

shintasmith replied:
Believe me, I tried. I thought I could do this:
```protected ByteBuffer toByteBuffer(<T extends Rollup> rollup)``` in the interface class. But it gave me a compile error when I tried to implement that with the specific Rollup subclass. The error is something like the method is not the same "erasure".

Any suggestions?

shintasmith replied:
I was able to do it with the ```fromByteBuffer()``` since it's the return type.

ChandraAddala replied:
I guess we can look into it later.

shintasmith commented:
@usnavi @ChandraAddala @VinnyQ @izrik @drenalin23 I think I am done updating this PR. I incorporated all your feedback. I also fixed the failing ```testLocatorWritten``` test in Travis.

Please let me know if you have additional comments/suggestions.

ChandraAddala commented:
Looks good.

shintasmith commented:
mergy mergy
